chore(website): Blog: What is the modern data stack (#9139)
cloudquery/cloudquery: website/pages/blog/what-is-the-modern-data-stack.mdx

---
title: "What is The Modern Data Stack?"
tag: engineering
date: 2023/03/16
description: >-

>   Past, Present and future of the modern data stack?

Reviewer: ```suggestion
  Past, Present and future of the modern data stack
```

author: yevgenypats
---

import { BlogHeader } from "../../components/BlogHeader"

<BlogHeader/>

> **Note**: The data world has become a complex landscape, full of acronyms and buzzwords. In this post we will give a birds-eye view and demystify all the terminology.

Reviewer: ```suggestion
The data world has become a complex landscape, full of acronyms and buzzwords. In this post we will give a birds-eye view of the modern data stack and demystify some of the terminology.
```

The modern data stack (MDS) is a set of tools and technologies that have emerged in recent years to help businesses collect, store, process, analyze, and visualize data more effectively. In contrast to the traditional data stack (TDS), the modern data stack makes use of cloud technologies to remove the burden, complexity and risk of managing a performant on-premise data system.

> In this blog, we will explore the key components of the modern data stack, popular tools and technologies, and use cases they unlock, such as Customer-Data-Platform (CDP), "Infrastructure-Data-Platform" (IDP), and many other custom use cases that involve the above processes.

Reviewer: ```suggestion
In this blog, we will explore the key components of the modern data stack, popular tools and technologies, and use cases they unlock, such as Customer Data Platform (CDP), Infrastructure Data Platform (IDP), CSPM (Cloud Security Posture Management) and many other acronyms and custom use cases that involve the above processes.
```

The modern data stack started gaining popularity in the early 2010s, enabled by cloud computing and OLAP data warehouse and data lake technologies such as Redshift, BigQuery, Snowflake, and more recent innovations like ClickHouse and in-memory OLAP technologies like DuckDB. Even though it has been around for a while, the popularity of this approach has grown significantly, and we are seeing a lot of innovation in every part of the stack.

The introduction of data warehouses and data lakes also made it possible to shift Extract-Transform-Load (ELT) workloads to Extract-Load-Transform (ELT). This approach moves transformation to the data warehouse and data lake which gives much more flexibility to teams working with the data without the need to ingest the data again and running compute intensive operations.

# Architecture

![The Modern Data Stack Architecture](/images/blog/what_is_the_modern_data_stack/architecture.png)

**Note:** There are many more vendors at each component of the stack and we chose some of them in no particular order (apart from CloudQuery given our bias :)).

## Data Sources

This refers to the data we want to ingest and involves different types of sources:

> - **Databases -** Companies have many databases for different workloads. Since these databases are not optimized for analytical workloads, it is often necessary to ingest data from production databases to analytical databases where you can run complex queries without impacting the production system. Additionally, there have been many advances around Change Data Capture ([CDC](https://www.cloudquery.io/blog/postgres-cdc-to-any-destination)) for databases, allowing you to ingest data in real-time from databases. This enables efficient, near-real time syncing between source databases and destinations.

Reviewer: ```suggestion
- **Databases**: Companies have many databases for different workloads. Since these databases are not optimized for analytical workloads, it is often necessary to ingest data from production databases to analytical databases where you can run complex queries without impacting the production system. Additionally, there have been many advances around Change Data Capture ([CDC](https://www.cloudquery.io/blog/postgres-cdc-to-any-destination)) for databases, allowing you to ingest data in real-time from databases. This enables efficient, near-real time syncing between source databases and destinations.
```

> - **APIs -** Most SaaS products expose their data via APIs. So, data for anything from marketing or finance products such as Salesforce and Stripe, to infrastructure platforms such as AWS, GCP, Azure can be loaded.**

Reviewer: ```suggestion
- **APIs**: Most SaaS products expose their data via APIs. So, data for anything from marketing or finance products such as Salesforce and Stripe, to infrastructure platforms such as AWS, GCP, Azure can be loaded.**
```

> - **Event Streams -** This is another popular source type where you might want to subscribe to an event queue and send it to an analytical destination.

Reviewer: ```suggestion
- **Event Streams**: This is another popular source type where you might want to subscribe to an event queue and send it to an analytical destination.
```
